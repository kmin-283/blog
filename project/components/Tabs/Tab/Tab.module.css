.tab {
    background-color: var(--main-theme-color);
    position: relative;
    z-index: 1;
    padding: 10px;
    list-style: none;
    text-align: center;
    cursor: pointer;
    overflow: hidden;
    width: 100%;
    font-weight: 500;
    color: white;
    border-radius: 5px 5px 0 0;
}

.tab:before {
    background-color: rgba(255, 255, 255, 0.3);
    position: absolute;
    top: 0;
    left: 0;
    z-index: -1;
    content: "";
    width: 100%;
    height: 120%;
    transition-duration: .25s;
    -webkit-transform: translateY(100%);
    transform: translateY(100%);
    border-radius: 5px 5px 0 0;
}

.tab:hover:before {
    -webkit-transform: translateY(70%);
    transform: translateY(70%);
}

.active {
    color: var(--main-theme-color);
    font-weight: 600;
}

.active:before, .active:hover:before {
    transition-duration: .5s;
    background-color: white;
    -webkit-transform: translateY(0);
    transform: translateY(0);
}