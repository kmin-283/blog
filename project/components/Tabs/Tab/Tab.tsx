import React, {FC} from 'react';
import styles from "./Tab.module.css";

export interface TabProps {
  tabId: string;
  active: boolean;
}

const Tab: FC<TabProps> = ({active, tabId, children}) => {
  return (
    <li className={`${styles.tab} ${active ? styles.active : ""}`} role="tab" aria-selected={active} aria-controls={tabId}>
      {children}
    </li>
  );
};

export default Tab;