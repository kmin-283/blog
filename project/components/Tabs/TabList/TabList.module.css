.tabList {
    display: flex;
    height: 80px;
    margin: 0;
    justify-content: space-evenly;
    align-items: center;
    border: 1px solid var(--border-line-color);
    background-color: #ad94fa;
}