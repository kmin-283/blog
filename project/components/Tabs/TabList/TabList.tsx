import React, {FC} from 'react';
import styles from './TabList.module.css';

const TabList: FC = ({children}) =>  {
  return (
    <ul className={styles.tabList} role="tablist" aria-label="Entertainment">
      {children}
    </ul>
  );
};

export default TabList;