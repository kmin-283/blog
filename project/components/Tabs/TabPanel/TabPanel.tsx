import React, {FC} from 'react';

interface TabPanelProps {
  isHidden: boolean;
  tabId: string;
}

const TabPanel: FC = ({ children, ...props}) => {
  const { isHidden, tabId } = props as TabPanelProps;
  return (
    <div role="tabpanel" aria-labelledby={tabId} hidden={isHidden}>
      {children}
    </div>
  );
};

export default TabPanel;