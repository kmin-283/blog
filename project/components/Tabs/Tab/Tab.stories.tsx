import React from 'react';
import { Story, Meta } from '@storybook/react';
import Tab from "@/components/Tabs/Tab/Tab";
import {TabProps} from "@/components/Tabs/Tab/Tab";

export default {
  title: 'Components/Tabs/Tab',
  component: Tab,
} as Meta;

const Template: Story<TabProps> = ({active, tabId}) => {
  return (
    <div style={{width:150, height:50}}>
      <Tab tabId={tabId} active={active}>게시된 글</Tab>
    </div>
  );
};

export const BasicTab = Template.bind({});
BasicTab.args = {
  /*👇 The args you need here will depend on your component */
  active: true,
  tabId: "published"
};