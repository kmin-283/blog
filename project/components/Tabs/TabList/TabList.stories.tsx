import React from 'react';
import { Story, Meta } from '@storybook/react';
import TabList from '@/components/Tabs/TabList/TabList';
import Tab from '@/components/Tabs/Tab/Tab';

export default {
  title: 'Components/Tabs/TabList',
  component: TabList,
  argTypes: {
    active: {
      options: ['publish', 'draft', 'test'],
      control: { type: 'radio'}
    }
  }
} as Meta;

const Template: Story = ({active}) => {
  return (
    <div style={{height:300}}>
      <TabList>
        <Tab tabId={"published"} active={active === 'publish'}>게시된 글</Tab>
        <Tab tabId={"drafted"} active={active === 'draft'}>임시 저장 글</Tab>
        <Tab tabId={"tested"} active={active === 'test'}>테스트 글</Tab>
      </TabList>
    </div>
  );
};

export const BasicTabList = Template.bind({});
BasicTabList.args = {
  active: 'publish'
};