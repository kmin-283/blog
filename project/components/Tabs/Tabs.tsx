import React, {FC, useState} from "react";
import styles from "./Tabs.module.css";
import TabList from "@/components/Tabs/TabList/TabList";
import TabPanel from "@/components/Tabs/TabPanel/TabPanel";

interface TabsProps {
  tabIds: string[];
}

const Tabs: FC<TabsProps> = ({children, tabIds}) => {
  const [tabId, setTabId] = useState<string>(tabIds[0]);
  const clickTab = (id: string) => () => {
    setTabId(id);
  };
  let tabIndex = 0;
  const childrenWithProps = React.Children.map(children, (child)=>{
    if (Object.getOwnPropertyDescriptor(child,'type')?.value.name === TabList.name && React.isValidElement(child)) {
      const tabs = React.Children.map(child.props.children, (grandChild) => {
        return React.cloneElement(grandChild, {
          active: tabId === tabIds[tabIndex],
          tabId: tabIds[tabIndex++],
          onClick: clickTab});
      });
      return React.cloneElement(child, {}, tabs);
    }

    if (Object.getOwnPropertyDescriptor(child,'type')?.value.name === TabPanel.name && React.isValidElement(child)) {
      if (tabIndex === tabIds.length) {
        tabIndex = 0;
      }
      return React.cloneElement(child, {
        isHidden: tabId !== tabIds[tabIndex],
        tabId: tabIds[tabIndex++],
      });
    }
    return child;
  });
  
  return (
    <div className={styles.tabs}>
      {childrenWithProps}
    </div>
  );
};

export default Tabs;
