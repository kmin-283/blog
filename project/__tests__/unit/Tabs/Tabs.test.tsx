import React from 'react';
import Tabs from '@/components/Tabs/Tabs';
import Tab from '@/components/Tabs/Tab/Tab';
import {render, cleanup} from '@testing-library/react';
import TabPanel from "@/components/Tabs/TabPanel/TabPanel";
import TabList from "@/components/Tabs/TabList/TabList";
import '@testing-library/jest-dom/extend-expect';

describe('Tabs 컴포넌트', () => {
  afterAll(cleanup);
  test('tabList를 가져야 한다', () => {
    const {getByRole} = render(<Tabs>
      <TabList>
        <Tab tabId={"published"} active={true}>게시된 글</Tab>
        <Tab tabId={"drafted"} active={false}>임시 저장 글</Tab>
      </TabList>
      <TabPanel tabId={"published"} isHidden={false}>
        <p>게시된 글 입니다</p>
      </TabPanel>
      <TabPanel tabId={"drafted"} isHidden={true}>
        <p>임시 저장 글 입니다</p>
      </TabPanel>
    </Tabs>);

    const tabList = getByRole('tablist');
    expect(tabList).toBeInTheDocument();
  });

  test('tabPanel을 가져야 한다', () => {
    const {getAllByRole} = render(
      <Tabs>
        <TabList>
          <Tab tabId={"published"} active={true}>게시된 글</Tab>
          <Tab tabId={"drafted"} active={false}>임시 저장 글</Tab>
          <Tab tabId={"tested"} active={false}>테스트 글</Tab>
        </TabList>
        <TabPanel tabId={"published"} isHidden={false}>
          <p>게시된 글 내용입니다</p>
        </TabPanel>
        <TabPanel tabId={"drafted"} isHidden={true}>
          <p>임시 저장 글 내용입니다</p>
        </TabPanel>
        <TabPanel tabId={"tested"} isHidden={true}>
          <p>테스트 글 내용입니다</p>
        </TabPanel>
      </Tabs>);

    // isHidden=true인 tabPanel은 보이지 않으므로 길이는 1이 된다.
    const tabPanels = getAllByRole('tabpanel');
    expect(tabPanels).toHaveLength(1);
    expect(tabPanels[0]).toBeInTheDocument();
  });
});