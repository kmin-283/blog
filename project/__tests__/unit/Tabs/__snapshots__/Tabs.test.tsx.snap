// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs 컴포넌트 렌더링 1`] = `
<div
  className="tabs"
>
  <ul
    aria-label="Entertainment"
    className="tabList"
    role="tablist"
  >
    <li
      aria-controls="published-tab"
      aria-selected={true}
      className="tab active"
      onClick={[Function]}
      role="tab"
    >
      게시된 글
    </li>
    <li
      aria-controls="drafted-tab"
      aria-selected={false}
      className="tab "
      onClick={[Function]}
      role="tab"
    >
      임시 저장 글
    </li>
  </ul>
  <div
    aria-labelledby="published"
    className="tabPanel"
    hidden={false}
    role="tabpanel"
  >
    <p>
      게시된 글 입니다
    </p>
  </div>
  <div
    aria-labelledby="drafted"
    className="tabPanel"
    hidden={true}
    role="tabpanel"
  >
    <p>
      임시 저장 글 입니다
    </p>
  </div>
</div>
`;
